sycl/test-e2e/WorkGroupMemory/swap_test_using_work_group_memory.cpp

```diff
@@ -0,0 +1,325 @@
+// RUN: %{build} -o %t.out
+// RUN: %{run} %t.out
+
+#include <cassert>
+#include <cstring>
+#include <sycl/detail/core.hpp>
+#include <sycl/ext/oneapi/experimental/work_group_memory.hpp>
+
+namespace syclexp = sycl::ext::oneapi::experimental;
+
+// This test performs a swap of two scalars/arrays inside a kernel using a
+// work_group_memory object as a temporary buffer. The test is done for scalar
+// types and bounded arrays. After the kernel finishes, it is verified on the
+// host side that the swap worked.
+
+// One important note is that for unbounded arrays, the feature is unstable.
+// Specifically, the code may or may not compile when kernels reference work
+// group memory objects that have been constructed with the unbounded array
+// type. This is due to a limitation of SPIRV where it does not allow arrays of
+// length zero. For example, an unbounded array may be translated to an array of
+// length zero in LLVM IR and during the LLVM IR -> SPIRV translation phase, the
+// translator rejects all arrays of length zero because they are invalid
+// constructs in SPIRV. As a result of this, unbounded arrays do not appear in
+// this test. They do appear in the sanity test though in this directory because
+// there the unbounded arrays are used with concrete subscript indices which
+// seems to work, for now at least.
```

Comment:
This comment seems concerning.  I don't understand the problem, though.  The implementation of `work_group_memory` is just a pointer.  The unbounded array is only used as the type of the template parameter.  Where does the zero-length array come from?

Reply:
I'll add a comment in the ticket.

Reply:
Instead of adding this comment, a test (you can make it separate `.cpp` file) should be added and `XFAIL`-ed.
Minimal reproducer is:
```c++
SYCL_EXTERNAL void foo() {
  sycl::ext::oneapi::experimental::work_group_memory<int[]> mem;

  mem[3] = 42;
}
```

Produced LLVM IR:
```
; Function Attrs: convergent mustprogress noinline norecurse nounwind optnone
define dso_local spir_func void @_Z3foov() #0 !srcloc !6 !sycl_fixed_targets !7 {
entry:
  %mem = alloca %"class.sycl::_V1::ext::oneapi::experimental::work_group_memory", align 8
  %mem.ascast = addrspacecast ptr %mem to ptr addrspace(4)
  call spir_func void @_ZN4sycl3_V13ext6oneapi12experimental17work_group_memoryIA_iNS3_10propertiesISt5tupleIJEEEEEC2Ev(ptr addrspace(4) noundef align 8 dereferenceable_or_null(24) %mem.ascast) #2
  %call = call spir_func noundef align 4 ptr addrspace(4) @_ZNK4sycl3_V13ext6oneapi12experimental17work_group_memoryIA_iNS3_10propertiesISt5tupleIJEEEEEcvRS5_Ev(ptr addrspace(4) noundef align 8 dereferenceable_or_null(24) %mem.ascast) #2
  %arrayidx = getelementptr inbounds [0 x i32], ptr addrspace(4) %call, i64 0, i64 3
  store i32 42, ptr addrspace(4) %arrayidx, align 4
  ret void
}
```

`sycl::_V1::ext::oneapi::experimental::work_group_memory<int [], sycl::_V1::ext::oneapi::experimental::properties<std::tuple<> > >::operator int (&) []() const` returns a reference to an unbounded array (if I read that correctly) and it seems like those aren't really pointers at LLVM IR level, but instead they are represented as zero-sized arrays.

My first guess is that fixing that at SPIR-V level would either be hacky or expensive (new extension + support for it in backends), so I think that we should investigate if we can somehow tweak generated LLVM IR instead. Perhaps we can also change the API somehow to avoid such generating such construct, but I don't know if that's possible

Reply:
Note that the example you give above is an invalid use of `work_group_memory`:

```
SYCL_EXTERNAL void foo() {
  sycl::ext::oneapi::experimental::work_group_memory<int[]> mem;

  mem[3] = 42;
}
```

The `work_group_memory` type can only be used to pass a parameter to a kernel.  It cannot be used to statically allocate work-group memory inside of a kernel.

Reply:
> Note that the example you give above is an invalid use of `work_group_memory`:

Yeah, I was simply trying to generate an LLVM IR example which is problematic, `mem` can be transformed to be `foo`'s argument and the issue will still be there:

```
; Function Attrs: convergent mustprogress noinline norecurse nounwind optnone
define dso_local spir_func void @_Z3fooN4sycl3_V13ext6oneapi12experimental17work_group_memoryIA_iNS3_10propertiesISt5tupleIJEEEEEE(ptr noundef byval(%"class.sycl::_V1::ext::oneapi::experimental::work_group_memory") align 8 %mem) #0 !srcloc !6 !sycl_fixed_targets !7 {
entry:
  %mem.ascast = addrspacecast ptr %mem to ptr addrspace(4)
  %call = call spir_func noundef align 4 ptr addrspace(4) @_ZNK4sycl3_V13ext6oneapi12experimental17work_group_memoryIA_iNS3_10propertiesISt5tupleIJEEEEEcvRS5_Ev(ptr addrspace(4) noundef align 8 dereferenceable_or_null(24) %mem.ascast) #2
  %arrayidx = getelementptr inbounds [0 x i32], ptr addrspace(4) %call, i64 0, i64 3
  store i32 42, ptr addrspace(4) %arrayidx, align 4
  ret void
}
```